File: Data Structures & Algorithms/valid-tree/submission-2.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        int[] parents = new int[n];
        Arrays.fill(parents, -1);

        int[] ranks = new int[n];
        Arrays.fill(ranks, 1);

        for (int[] edge : edges) {
            int node = edge[0];
            int nbr = edge[1];

            int p1 = find(parents, node);
            int p2 = find(parents, nbr);

            System.out.println("Node: " + node);
            System.out.println("parent: " + p1);

            System.out.println("Nbr: " + nbr);
            System.out.println("parent: " + p2);
            System.out.println();

            if (p1 != p2) {
                if (ranks[p1] < ranks[p2]) {
                    ranks[p2] += ranks[p1];
                    parents[p1] = p2;
                } else {
                    ranks[p1] += ranks[p2];
                    parents[p2] = p1;
                }
            } else{
                return false;
            }
        }
        return true;
    }

    public int find(int[] parents, int i) {
        if (parents[i] == -1) {
            return i;
        }
        int p = find(parents, parents[i]);
        parents[i] = p;
        return p;
    }
}

File: Data Structures & Algorithms/redundant-connection/submission-2.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges) + 1
        parents = [-1] * n
        ranks = [1] * n


        def find(index):
            if parents[index] == -1:
                return index

            p = find(parents[index])
            parents[index] = p
            return p

        res = []
        for edge in edges:
            
            node = edge[0]
            nbr  = edge[1]

            p1 = find(node)
            p2 = find(nbr)

            if p1 != p2:
                if ranks[p1] < ranks[p2]:
                    ranks[p2] += ranks[p1]
                    parents[p1] = p2

                else:
                    ranks[p1] += ranks[p2]
                    parents[p2] = p1

            if p1 == p2:
                return edge
        return res
